godot-macros/src/godot_api.rs

```diff
@@ -301,48 +305,61 @@ fn extract_attributes<T>(
-        let attr_name = attr
-            .get_single_path_segment()
-            .expect("get_single_path_segment");
-
-        let new_found = match attr_name {
-            name if name == "func" => Some(BoundAttr {
-                attr_name: attr_name.clone(),
-                index,
-                ty: BoundAttrType::Func(attr.value.clone()),
-            }),
-            name if name == "signal" => {
-                // TODO once parameters are supported, this should probably be moved to the struct definition
-                // E.g. a zero-sized type Signal<(i32, String)> with a provided emit(i32, String) method
-                // This could even be made public (callable on the struct obj itself)
-                Some(BoundAttr {
-                    attr_name: attr_name.clone(),
-                    index,
-                    ty: BoundAttrType::Signal(attr.value.clone()),
-                })
-            }
-            name if name == "constant" => Some(BoundAttr {
-                attr_name: attr_name.clone(),
-                index,
-                ty: BoundAttrType::Const(attr.value.clone()),
-            }),
-            _ => None,
-        };
+    const FUNC: &str = "func";
+    const SIGNAL: &str = "signal";
+    const CONSTANT: &str = "constant";
+
+    let mut name_and_parser = None;
 
-        // Validate at most 1 attribute
-        if found.is_some() && new_found.is_some() {
-            bail!(
-                &error_scope,
-                "at most one #[func], #[signal], or #[constant] attribute per declaration allowed",
-            )?;
+    let mut found_attr_count = 0;
+    for name in [FUNC, SIGNAL, CONSTANT] {
+        if let Some(valid_parser) = KvParser::parse(attributes, name)? {
+            found_attr_count += 1;
+            name_and_parser = Some((name, valid_parser));
         }
+    }
 
-        found = new_found;
+    if found_attr_count > 1 {
+        bail!(
+            &error_scope,
+            "at most one #[func], #[signal], or #[constant] attribute per declaration allowed",
+        )?;
     }
 
-    Ok(found)
+    let Some((name, mut parser)) = name_and_parser else {
+        return Ok(None);
+    };
+
+    let index = parser.index();
+
+    let attr = match name {
+        FUNC => {
+            // TODO you-win (August 8, 2023): handle default values here as well?
+
+            let rename = parser.handle_expr("rename")?.map(|value| value.to_string());
+
+            Some(BoundAttr {
+                attr_name: name,
+                index,
+                ty: BoundAttrType::Func { rename },
+            })
+        }
+        SIGNAL => Some(BoundAttr {
+            attr_name: name,
+            index,
+            ty: BoundAttrType::Signal,
+        }),
+        CONSTANT => Some(BoundAttr {
+            attr_name: name,
+            index,
+            ty: BoundAttrType::Const,
+        }),
+        _ => unreachable!(),
+    };
+
+    parser.finish()?;
```

Comment:
Regarding the flow, what about [my comment here](https://github.com/godot-rust/gdext/pull/376#discussion_r1288113591)?

Does defining separate parsers only not work because of `#[allow(clippy::uninlined_format_args)]`? If possible we should fix that clippy lint, not suppress it.

Reply:
Because only one attribute is supported at a time, the initial proposed flow introduces redundant `if let Some(...)` blocks that need to be checked every time + bailing out early. The problem isn't separate parsers, just that separate parsers aren't needed or wanted in this instance.

Reply:
It's just that looping over 3 constants is pointless, if the whole loop body is matching that very constant and dispatching to specific code. There's no business logic that can be reused inside the loop, except maybe `finish()`.

If you're worried about duplicates, you can do something like this:
```rs
let func = KvParser::parse(attributes, "func")?;
let signal = KvParser::parse(attributes, "signal")?;
let constant = KvParser::parse(attributes, "constant")?;

if [func, signal, constant]
    .filter(|p| p.is_some()).count() > 1 {
    // bail...
}

// now handle each case
```
or:
```rs
let mut ty: Option<BoundAttrType> = None;

if let Some(func) = KvParser::parse(attributes, "func")? {
   // handle #[func]...
   ty = ...;
   parser.finish()?;
}

if let signal = KvParser::parse(attributes, "signal")? {
   if ty.is_some() {
       // bail...
   }

   // handle #[constant]...
   ty = ...;
   parser.finish()?;
}

if let signal = KvParser::parse(attributes, "constant")? {
   if ty.is_some() {
       // bail...
   }

   // handle #[constant]...
   ty = ...;
   parser.finish()?;
}
```

There may be a tiny bit of code duplication, but it's much easier to understand, you just handle each case sequentially.

Reply:
> allows for the original `match` logic to be retained, which makes it clear that only 1 option is valid and only 1 parser should be handled

This is already clear with the validation, in both your and my case. And there's a lot of code just to make this flow work, which aren't needed in my approach (the constants, the loop, match, unreachable).

Anyway, I'm repeating myself and I don't think this is important enough to spend more time on discussing 🙂 you can leave it if you like, maybe we find a better pattern for exclusive `KvParser`s in the future.